psh/test-date.py

```diff
@@ -0,0 +1,197 @@
+    psh.assert_cmd(p, "date", expected=DFLT_DATE_REG, is_regex=True)
+
+    # Incorrect FORMAT passed when printing date
+    psh.assert_cmd(p, "date +operand1", expected="operand1")
+    psh.assert_cmd(p, "date", expected=DFLT_DATE_REG, is_regex=True)
+
+    # too many arguments passed to print, should print first redundant
+    psh.assert_cmd(p, "date operand1 operand2 operand3", expected=unrec_arg("operand2"), result="fail")
+    psh.assert_cmd(p, "date", expected=DFLT_DATE_REG, is_regex=True)
+
+    # nonexistent format '%k' passed to print
+    psh.assert_cmd(p, "date +%Y%k%Y", expected=r"\b(19|20)\d{2}%k(19|20)\d{2}\r+\n", is_regex=True)
+    psh.assert_cmd(p, "date", expected=DFLT_DATE_REG, is_regex=True)
+
+
+def test_incorrect_datewrite(p):
+    """Test incorrect commandlines for setting date (no edge cases)"""
+
+    # No argument passed
+    psh.assert_cmd(p, "date -s", expected="date: option requires an argument -- s", result="fail")
+    psh.assert_cmd(p, "date", expected=DFLT_DATE_REG, is_regex=True)
+
+    # invalid time value
+    psh.assert_cmd(p, "date -s 123456789operand1", expected=invalid_date("123456789operand1"), result="fail")
+
+    # too many arguments
+    psh.assert_cmd(p, "date -s 1600000000 +format operand1", expected=unrec_arg("operand1"), result="fail")
+    psh.assert_cmd(p, "date", expected=DFLT_DATE_REG, is_regex=True)
+
+    # no integer number
+    psh.assert_cmd(p, "date -s @1600000000.1234567890", expected=invalid_date("@1600000000.1234567890"), result="fail")
+    psh.assert_cmd(p, "date", expected=DFLT_DATE_REG, is_regex=True)
+
+
+def test_incorrect_dateparse(p):
+    # No argument passed
+    psh.assert_cmd(p, "date -d", expected="date: option requires an argument -- d", result="fail")
+    psh.assert_cmd(p, "date", expected=DFLT_DATE_REG, is_regex=True)
+
+    # too many arguments
+    psh.assert_cmd(p, "date -d @1600000000 +format operand1", expected=unrec_arg("operand1"), result="fail")
+    psh.assert_cmd(p, "date", expected=DFLT_DATE_REG, is_regex=True)
+
+    # invalid time value
+    psh.assert_cmd(p, "date -d @1600000000operand1", expected=invalid_date("@1600000000operand1"), result="fail")
+    psh.assert_cmd(p, "date", expected=DFLT_DATE_REG, is_regex=True)
+
+    # no '@' sign
+    psh.assert_cmd(p, "date -d 1600000000", expected=invalid_date("1600000000"), result="fail")
+    psh.assert_cmd(p, "date", expected=DFLT_DATE_REG, is_regex=True)
+
+    # no integer number
+    psh.assert_cmd(p, "date -d @1600000000.1234567890", expected=invalid_date("@1600000000.1234567890"), result="fail")
+    psh.assert_cmd(p, "date", expected=DFLT_DATE_REG, is_regex=True)
+
+
+def test_edges(p):
+    """Test edge cases for epoch value passed to `date -s epoch"""
+
+    # maximum value for int32
+    psh.assert_cmd(
+        p,
+        "date -s @2147483647",
+        expected=r"Tue, 19 Jan (?:38|2038),? \d{2}:\d{2}:\d{2}(?:\w{3}|\w{4})?\r+\n",
```

Comment:
Please place it in one variable and use exact value: `Tue, 19 Jan 38 03:14:07`, we can assume that time zones won't be added and an hour won't change (alternatively `GMT` will be printed).